### PR TITLE
Implement NorMuon

### PR DESCRIPTION
**Pull request by zichongli5**
### 🚀 Add NorMuon Optimizer (Neuron-wise Normalized Muon)

This PR adds **[NorMuon](https://arxiv.org/pdf/2510.05491)**, a variant of the Muon optimizer that introduces **neuron-wise normalization** to improve stability and convergence efficiency.

#### Results
Built on the current master branch (excluding recent PRs), NorMuon reaches the **3.28 loss cutoff in 1700 steps**, 50 steps faster than Muon with identical per-step time.
 With 1700 steps schedule, Muon achieves a validation loss of **3.2859**, while NorMuon achieves **3.2784 / 3.2795 / 3.2793** across three runs.

#### Hyperparameters
We keep the same hyperparameters and have not tuned them for NorMuon. The performance might be further improved with tuning.

#### Notes
This implementation includes two slight modifications from the paper version:

1. The update is rescaled to match the norm of the pre-normalized update.
2. In the algorithm from the paper (line 7), we average the second-order momentum across columns. In this implementation we average across columns if m ≥ n and across rows otherwise, which performs slightly better.

### CODE REVIEW

**Comment by ClassicLarry** [comment]
Awesome, this improvement looks quite substantial! The code update also looks very compact for the size of impact. I am going to merge in the other PRs this week on the 124M track, which are now 25s faster than this one on its own. However, I'm quite confident if these same changes get applied to the latest PR this will show a notable improvement, with maybe some minor tuning to account for change in batch size. Please see rule 2 in readme for details on p value for submission.

**Comment by varunneal** [comment]
@zichongli5 it looks like this implementation is pretty similar to adafactor, but instead of computing a full 2d mesh, it just computes it along one dimension (e.g. either along rows or cols). I'm wondering
- why not apply the second moment along both rows *and* cols?
- why not apply the second moment prior to the zero power function?

**Comment by zichongli5** [comment]
> @zichongli5 it looks like this implementation is pretty similar to adafactor, but instead of computing a full 2d mesh, it just computes it along one dimension (e.g. either along rows or cols). I'm wondering
> 
> * why not apply the second moment along both rows _and_ cols?
> * why not apply the second moment prior to the zero power function?

Yes — this follows the similar idea as adafactor/adammini.

For the first question: That’s a valid extension, though I kept it 1-D for simplicity and minimal state; applying both might give a small gain.

For the second question: We tested applying second moment before ns5 in the paper (see Fig. 3) and found very similar results.

**Comment by ClassicLarry** [comment]
Closing as #144, an updated version of this PR, has been merged.
